assert/jsonmatch/matches_by.go

```diff
@@ -0,0 +1,134 @@
+package jsonmatch
+
+import (
+	"errors"
+)
+
+type (
+	// reprepsents a node with a value in a json object
+	Node struct {
+		kind  Kind
+		value interface{}
+	}
+
+	// an enum of the kinds of nodes in a json object
+	Kind int
+
+	// a map of matchers with a function to determine if a value matches
+	// the key of the matchers corresponds to the value passed into
+	// the function
+	Matchers map[string]func(interface{}) bool
+)
+
+const (
+	Object Kind = iota
+	Array
+	String
+	Number
+	Bool
+	Null
+)
+
+var ErrInvalidType = errors.New("invalid data type in json object")
+
+func MatchesBy(expected, actual interface{}, matchers Matchers) bool {
+	expectedNode := New(expected)
+	actualNode := New(actual)
+
+	return expectedNode.matchesBy(actualNode, matchers)
+}
+
+func New(v interface{}) Node {
+	n := Node{}
+
+	switch v := v.(type) {
+	case map[string]interface{}:
+		n.kind = Object
+		val := make(map[string]Node)
+
+		for k, childVal := range v {
+			val[k] = New(childVal)
+		}
+
+		n.value = val
+	case []interface{}:
+		n.kind = Array
+		val := make([]Node, 0)
+		for _, childVal := range v {
+			val = append(val, New(childVal))
+		}
+		n.value = val
+	case string:
+		n.kind = String
+		n.value = v
+	case float64:
+		n.kind = Number
+		n.value = v
```

Comment:
What about json.Number?

Reply:
instead of float64 for instance?

Reply:
I'm not opposed to it, but what benefit would it provide?

Reply:
I'm sorry if I didn't explain it well.

Of course, the float64 case is needed.

My point was about the fact, and maybe I'm wrong, this function could receive data that was, or will be, decoded with UseNumber

https://pkg.go.dev/encoding/json#Decoder.UseNumber

But maybe I'm wrong and it cannot happen

Reply:
ahh, i see.  I think that would have been caught by this test here: https://github.com/stretchr/testify/pull/1690/files#diff-7190065248695df477329c4cada2d71f323ab7940ecffff2d3d5fa43caf79427R2454